### PR TITLE
MRG: Add projection of EEG electrodes

### PR DESCRIPTION
**Pull request by Eric89GXL**
This allows visualizing the EEG electrodes projected onto the scalp.

Also gets the machinery in place for #3700.

Ready for review/merge from my end.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
CircleCI output:

https://2923-1301584-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/auto_examples/visualization/plot_eeg_on_scalp.html

File: mne/viz/_3d.py [hunk 1]
```diff
@@ -335,98 +367,100 @@ def plot_trans(info, trans='auto', subject=None, subjects_dir=None,
     elif not isinstance(trans, dict):
         raise TypeError('trans must be str, dict, or None')
     head_mri_t = _ensure_trans(trans, 'head', 'mri')
+    dev_head_t = info['dev_head_t']
     del trans
 
+    # Figure out our transformations
+    if coord_frame == 'meg':
+        head_trans = invert_transform(dev_head_t)
+        meg_trans = Transform('meg', 'meg')
```
Comment (jaeilepp) [inline mne/viz/_3d.py]: Shouldn't there be a third argument? Can you add a test for this?

File: mne/viz/_3d.py [hunk 2]
```diff
@@ -312,15 +316,43 @@ def plot_trans(info, trans='auto', subject=None, subjects_dir=None,
...
+        if not isinstance(var, (list, tuple)) or \
+                not all(isinstance(x, string_types) for x in var):
+            raise TypeError('%s_sensors must be list or tuple of str, got %s'
+                            % (type(var),))
+    if not all(x in ('helmet', 'points') for x in meg_sensors):
+        raise ValueError('meg_sensors must only contain "helmet" and "points",'
+                         ' got %s' % (meg_sensors,))
+    if not all(x in ('original', 'projected') for x in eeg_sensors):
+        raise ValueError('eeg_sensors must only contain "original" and '
+                         '"projected", got %s' % (meg_sensors,))
```
Comment (jaeilepp) [inline mne/viz/_3d.py]: ``meg_sensors`` -> ``eeg_sensors``

File: mne/forward/_compute_forward.py [hunk 3]
```diff
@@ -234,27 +233,20 @@ def _bem_specify_els(bem, els, mults):
...
-            best = _find_nearest_tri_pt(inds, r, scalp['geom'], True)[2]
-            # Calculate a linear interpolation between the vertex values
-            tri = scalp['tris'][best]  # Get 3 vertex indices of closest tri
-            # Get coords of pt projected onto closest triangle
-            x, y, z = _triangle_coords(r, scalp['geom'], best)
-            w = elw * np.array([(1.0 - x - y), x, y])
-            amt = np.dot(w, bem['solution'][tri])
-            sol[k] += amt
+
+    # Operate on all intgration points for all electrodes (in MRI coords)
```
Comment (wmvanvliet) [inline mne/forward/_compute_forward.py]: intgration?

File: mne/viz/_3d.py [hunk 4]
```diff
@@ -312,15 +316,43 @@ def plot_trans(info, trans='auto', subject=None, subjects_dir=None,
...
+
     if not isinstance(info, Info):
         raise TypeError('info must be an instance of Info, got %s'
                         % type(info))
     if coord_frame not in ['head', 'meg', 'mri']:
         raise ValueError('coord_frame must be "head" or "meg"')
```
Comment (jaeilepp) [inline mne/viz/_3d.py]: ``'mri``` should be added to the error.

File: mne/viz/_3d.py [hunk 5]
```diff
@@ -335,98 +367,100 @@ def plot_trans(info, trans='auto', subject=None, subjects_dir=None,
...
+                           for k in pick_types(info, meg=False, eeg=True,
+                                               ref_meg=False)])
         if len(eeg_loc) > 0:
-            # Transform EEG electrodes from head coordinates if necessary
-            if coord_frame == 'meg':
-                eeg_loc = apply_trans(invert_transform(info['dev_head_t']),
-                                      eeg_loc)
-            elif coord_frame == 'mri':
-                eeg_loc = apply_trans(head_mri_t, eeg_loc)
+            eeg_loc = apply_trans(head_trans, eeg_loc)
```
Comment (jaeilepp) [inline mne/viz/_3d.py]: I would put this inside the ``if 'projected' in eeg_sensors:`` statement.

Comment (Eric89GXL) [inline mne/viz/_3d.py]: no, `eeg_loc` needs to get transformed to the proper coordinate frame regardless of whether or not `eegp_loc` will be used

**Comment by Eric89GXL** [comment]
Thanks for the look @jaeilepp, comments addressed

**Comment by agramfort** [comment]
the fact that projected sensors have the same red color makes it really had to see.

2 options:
- use a different color for projected sensors
- only allow one way for plotting. Eventually showing the two ways in the example?

thoughts?

**Comment by Eric89GXL** [comment]
I think it's nice to be able to see both. Are they all hard for you to see? For me they are hard to see in the case that they overlap, which is precisely when I don't need to see them both...

If you'd prefer I can change the colors to be more different, but they should be *somewhat* similar to differentiate them from, say, dig points, or any other type of points.

Another option would be to draw a line from the original to the spot on the scalp, but not sure how easy this is in Mayavi or how good it would look.

**Comment by agramfort** [comment]
whatever works. It's just maybe in static view in web page that it's not
really visible :(

**Comment by Eric89GXL** [comment]
Can you try it locally? That way you could see if it's better interactively, and easily make changes to see if you like something better. Might be faster than iterating over GitHub comments

File: mne/viz/_3d.py [hunk 6]
```diff
@@ -444,10 +480,10 @@ def plot_trans(info, trans='auto', subject=None, subjects_dir=None,
         mesh.data.cell_data.normals = None
         mlab.pipeline.surface(mesh, color=colors[key], opacity=alphas[key])
 
-    datas = (eeg_loc, hpi_loc, car_loc, ext_loc)
-    colors = ((1., 0., 0.), (0., 1., 0.), (1., 1., 0.), (1., 0.5, 0.))
-    alphas = (1.0, 0.5, 0.5, 0.25)
-    scales = (0.005, 0.015, 0.015, 0.0075)
+    datas = (eeg_loc, eegp_loc, hpi_loc, car_loc, ext_loc)
+    colors = ((1., 0, 0), (0.5, 0, 0), (0., 1, 0), (1., 1, 0), (1, 0.5, 0))
```
Comment (agramfort) [inline mne/viz/_3d.py]: try:

    colors = ((1., 0, 0), (0., 0.9, 1.), (0., 1, 0), (1., 1, 0), (1, 0.5, 0))

to me it's much nicer :)

Comment (Eric89GXL) [inline mne/viz/_3d.py]: I don't like that the "projected" is so dissimilar from the "original" red, but most of the similar colors are actually taken by other things. So I tweaked `(0, 0.9, 1)` to `(0, 0.5, 1)` and it looks more MNE-C like (and still good I think; ignore the blockiness/normal problems, that's my own mayavi build issue):

screenshot from 2016-10-28 19-13-00


Okay?

**Comment by agramfort** [comment]
thanks @Eric89GXL

**Pull request merged**